### ISSUE TITLE
set_seed() should take std::seed_seq instead of int (and other minor RNG issues)

### ISSUE THREAD

**Issue by patrikhuber** [tiny-dnn/tiny-dnn#285]
I believe [random_generator::set_seed(unsigned int seed)](https://github.com/tiny-dnn/tiny-dnn/blob/master/tiny_dnn/util/util.h#L84) should take a [`std::seed_seq`](http://en.cppreference.com/w/cpp/numeric/random/seed_seq) instead of of an `uint`. We should give the mt19937 the entropy it requires - it has quite a large state.
I believe for neural network training this might matter, but I don't have much experience in this regard.

And we should probably do the same for the [default construction](https://github.com/tiny-dnn/tiny-dnn/blob/master/tiny_dnn/util/util.h#L90) - currently it's seeded with just "1" (one int) which is not enough entropy for the mt19937. A deterministic seed is ok I think (it's useful for reproducibility / debugging) but probably we should use `std::seed_seq` here as well.

Also according to the MS Connect bug report, [this](https://github.com/tiny-dnn/tiny-dnn/blob/master/tiny_dnn/util/util.h#L88-L89) has been fixed in >=VS2013 so we could remove the note.

[This blog post](http://www.elbeno.com/blog/?p=1318) and the [explanations](http://www.elbeno.com/blog/?p=1325) provide a nice overview of the subject. Also [this](http://codereview.stackexchange.com/questions/109260/seed-stdmt19937-from-stdrandom-device).

**Comment by nyanp** [comment]
@patrikhuber 
Sorry for late reply, thank you for information! It's interesting topic for me.

I vote for determinstic seed for default construction. Reproduciblity is critical for industrial use, and  expected behavior for many users because caffe/tensorflow use fixed seed in default.

I have no idea about how the entropy affects the performance of neural networks, but taking ```std::seed_seq``` as an argument seems reasonable. I prefer to provide both ```set_seed(unsigned int)``` and ```set_seed(std::seed_seq)``` (I think the former is needed for casual/non-c++ expert user).

**Comment by bhack** [comment]
Caffe has https://github.com/bvlc/caffe/blob/master/src/caffe/common.cpp#L60

**Comment by patrikhuber** [comment]
@bhack: Interesting. I guess they are either not aware of proper RNG initialisation or it doesn't matter for the NN training at all - my guess is both are true :-)

Adding an overload to `set_seed()` that takes a `seed_seq`, and keeping the current one, sounds like the best to do. In fact if we ever want to find out, it would easily allow to compare the two approaches to find out if the RNG or its entropy has any significance in the training or convergence rate. My guess is that the quality of the randomness doesn't matter for the weights initialisation. But it's still good if we have a method to properly initialise the RNG.

**Comment by bhack** [comment]
Caffe use as default https://github.com/BVLC/caffe/blob/be163be0ea5befada208dbf0db29e6fa5811dc86/include/caffe/util/rng.hpp#L16

**Comment by bhack** [comment]
See also TF behavior http://stackoverflow.com/questions/36288235/how-to-get-stable-results-with-tensorflow-setting-random-seed

**Comment by patrikhuber** [comment]
Ah, I found a RNG constructor in Caffe sources that initialises with an int, maybe I looked in the wrong place. Now the question is what is `Caffe::rng_stream().generator()`. It's hard to browse the sources on GitHub, I don't have Caffe checked out :-)

Seems like TF uses just an `int` too. But that itself isn't saying much yet though, there are indeed RNGs that require much less entropy than `mt19937`, where an `int` might be enough to seed the full state. But probably TF uses an MT too.
If you're interested, http://www.pcg-random.org/ is also a great resource.

**Comment by bhack** [comment]
Another Caffe constructor use https://github.com/bvlc/caffe/blob/master/src/caffe/common.cpp#L23.
For Tensorflow random init and distributions see https://github.com/tensorflow/tensorflow/tree/master/tensorflow/core/lib/random

**Comment by patrikhuber** [comment]
I found the Caffe constructor you linked too, that one also just returns an int64.

**Comment by bhack** [comment]
See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/lib/random/random.cc

**Comment by patrikhuber** [comment]
Same there, I believe `device()` will return an `int` or something similar.

Anyway, this code is contrary to what @nyanp said:

> because caffe/tensorflow use fixed seed in default.

Maybe the function you linked is not actually used in their code.

**Comment by bhack** [comment]
Don't used fixed seed by default but could be set when you need reproducibility.

**Comment by Randl** [comment]
@patrikhuber @bhack @nyanp Is this still relevant?

**Comment by patrikhuber** [comment]
Well, I think I would initialise the mt19937 with a `seed_seq` to give it the entropy it requires, but as mentioned before I really don't know whether the quality of random numbers matters for deep learning training.

**Comment by Randl** [comment]
@patrikhuber Could you send PR if you think that it worth the effort?

**Comment by pliptor** [comment]
As far as I understand, MT19937ar has two types of initializing functions. 
(1). Takes a single long.
(2). Takes an array of long.

http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/MT2002/CODES/mt19937ar.c

You'll see that (1) feeds into an auxiliary PRNG to fill MT's entire internal state. Having both interfaces available to the user is good but I think it's unlikely convergence rate of a CNN is going to be affected by their choice. The second feeding type is mostly useful for applications that must to get more than 2^32 distinct seeds. This is not an easy subject though. MT initialization has evolved you can find some interesting stories in their web site. 

http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/MT2002/emt19937ar.html
